### PR TITLE
Restructuring the provenance log HTML building functionality to separate log entries by metadata field

### PR DESCRIPTION
**Pull request by jrgriffiniii**
Advances #675 by restructuring the provenance log HTML building functionality to ensure that individual field changes have users and timestamps associated with each field change:

image

### CODE REVIEW

**Comment by mccalluc** [comment]
I also wonder if this PR is refactoring more than is necessary. And, I'm worried about merge conflicts with
- #692
- #693

**Comment by jrgriffiniii** [comment]
I am sorry, I will look to port these proposed changes into https://github.com/pulibrary/pdc_describe/pull/692

**Comment by bess** [comment]
@jrgriffiniii Should this PR be closed in favor of #692 ?

File: app/models/work_activity.rb [hunk 1]
```diff
@@ -35,37 +39,58 @@ def users_referenced
...
+    if changes_event_type?
       metadata_changes_html
-    elsif activity_type == "FILE-CHANGES"
+    elsif file_changes_event_type?
       file_changes_html
     else
       comment_html
     end
   end
+  alias message_html to_html
```
[the remark this answers is not in the document]
Comment (mccalluc) [reply app/models/work_activity.rb:94]: I'm still not sure why this is needed.

File: app/models/work_activity.rb [hunk 2]
```diff
@@ -39,34 +45,82 @@ def users_referenced
...
+      created_at_time.strftime("%B %d, %Y %H:%M")
+    end
+
+    def event_timestamp_html
+      "#{event_timestamp} by " if system_event_type? || log_event_type?
+    end
+
+    # This is working
+    def comment_timestamp_html
+      "at #{event_timestamp}" if comment_event_type?
```
Comment (mccalluc) [inline app/models/work_activity.rb:112]: I'm not sure what `This is working` means. Should there also be a space at the start of the string? If there are going to be paired methods like this, would it be safer to make one `if` and the other `unless`, so we can be confident that precisely one is engaged?

File: app/models/work_activity.rb [hunk 3]
```diff
@@ -80,20 +134,69 @@ def file_changes_html
...
     def metadata_changes_html
       text = ""
       changes = JSON.parse(message)
+
       changes.keys.each do |field|
-        values = changes[field].map { |value| change_value_html(value) }.join
-        text += "<p><b>#{field}</b>: #{values}</p>"
+        change = changes[field]
+        mapped = change.map { |value| change_value_html(value) }
+        values = mapped.join
```
Comment (mccalluc) [inline app/models/work_activity.rb:150]: Should we in general avoid method chaining?

**Review by mccalluc — COMMENTED** [review]
There are things I'm uncomfortable with here, but I don't want to block.

File: app/models/work_activity.rb (hunk 3 above)
Comment (jrgriffiniii) [reply app/models/work_activity.rb:150]: I apologize, my approach has been to restructure these out of habit (I have attempting to follow the guidance of Sandi Metz: https://sandimetz.com/). I can revert these changes.

File: app/views/works/_work_activity.html.erb [hunk 4]
```diff
@@ -1,7 +1,3 @@
-<div class="activity-history-item activity-history-<%= event_type %>-item">
-  <div class="activity-history-<%= event_type %>-title">
-    <b><%= activity.created_by_user&.display_name_safe || "Unknown user outside the system" %></b>
-    <%= activity.created_at.time.strftime("%B %d, %Y %H:%M") %>
-  </div>
-  <span class="comment-html"><%= activity.message_html.gsub('{USER-PATH-PLACEHOLDER}', users_path).html_safe %></span>
+<div>
+  <%= activity.to_html.html_safe %>
 </div>
```
[the remark this answers is not in the document]
Comment (jrgriffiniii) [reply app/views/works/_work_activity.html.erb:3]: Excellent point! I can look to address this, thank you.

File: app/models/work_activity.rb [hunk 5]
```diff
@@ -35,37 +39,58 @@ def users_referenced
...
-      parsed_document = Kramdown::Document.new(text)
-      parsed_document.to_html
+    def event_html(children:)
+      <<-HTML
+<div class="activity-history-#{event_type}-title">
+    <b>#{created_by_user_html}</b>
+  #{created_at_html}
+  </div>
+<span class="comment-html">#{children}</span>
+    HTML
```
[the remark this answers is not in the document]
Comment (jrgriffiniii) [reply app/models/work_activity.rb:123]: I am sorry for this, and I have no strong opinions. However, I believe that I recall Blacklight (https://github.com/projectblacklight/blacklight) uses nested, recursive partial rendering, and this slows the responses *very* noticeably.

I was going to please propose two alternatives to any markup in the `models`:
- `Presenter` design pattern (https://www.rubyguides.com/2019/09/rails-patterns-presenter-service/)
- `Decorator` design pattern (https://www.rubyguides.com/2018/04/decorator-pattern-in-ruby/)

**Review by bess — APPROVED** [review]
We discussed this in stand up and decided to merge as is. :)

**Pull request merged**